---
title: "EGS003SALVEseq"
author: "Emanuelle Grody"
date: "2024-Summer"
output: html_document
---

This file is for processing SALVEseq reads


```{r, echo = FALSE}
source("~/SALVEseq/packages.R")
source("~/SALVEseq/functions.R")
```

# env and tat Analysis
## Previous analysis
### Loading in Seurat objs
```{r}

PL_env <- SeuratSALVEseq("/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/analysis/counts/run_count_PL_env/outs/filtered_feature_bc_matrix_scPathoQuant_bbmap/", 
                     "PL_env", 
                     output_dir = "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/analysis/Seurat/PL_env/", plots = TRUE)
PL_tat1 <- SeuratSALVEseq("/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/analysis/counts/run_count_PL_tat1/outs/filtered_feature_bc_matrix_scPathoQuant_bbmap/", 
                     "PL_tat1", 
                     output_dir = "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/analysis/Seurat/PL_tat1/", plots = TRUE)
PL_tat2 <- SeuratSALVEseq("/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/analysis/counts/run_count_PL_tat2/outs/filtered_feature_bc_matrix_scPathoQuant_bbmap/", 
                     "PL_tat2", 
                     output_dir = "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/analysis/Seurat/PL_tat2/", plots = TRUE)
```

Next we isolate the viral counts from our Seurat object.
```{r scPathoCountCellID}

output_dir = "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/analysis/Seurat/PL_env/"
PLenv_mac239 <- scPathoCountCellID(PL_env, "mac239")
write.csv(PLenv_mac239, paste0(output_dir, "PLenv_mac239expression.csv"))

output_dir = "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/analysis/Seurat/PL_tat1/"
PLtat1_mac239 <- scPathoCountCellID(PL_tat1, "mac239")
write.csv(PLtat1_mac239, paste0(output_dir, "PLtat1_mac239expression.csv"))

output_dir = "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/analysis/Seurat/PL_tat2/"
PLtat2_mac239 <- scPathoCountCellID(PL_tat2, "mac239")
write.csv(PLtat2_mac239, paste0(output_dir, "PLtat2_mac239expression.csv"))
```

These cellIDs are already whitelist filtered.
Interacting with the data: what does the mac239 expression look like?
```{r}
#How many cells have mac239 expression?
hist(PLenv_mac239$Count)
hist(PLtat1_mac239$Count)
hist(PLtat2_mac239$Count)

(sum(PLtat1_mac239$Count == 0) / nrow(PLtat1_mac239))

#How many reads are mac239 reads?
count_matrix <- GetAssayData(PL_tat1, slot = "counts")
cell_sums <- colSums(count_matrix)
PLtat1_allgenes <- data.frame(
    cellID = names(cell_sums),
    totalCounts = cell_sums)

PLtat1_ratio <- inner_join(PLtat1_mac239, PLtat1_allgenes, by = "cellID") %>% mutate(ratio = Count / totalCounts)
hist(PLtat1_ratio$ratio)
```

Currently no UMI cutoff set.


## Current analysis

### Loading Single Cell data
We load in the 10X data with viral counts and UMAP coordinates (see EGS002scRNAseq.Rm for the analysis that generated these datasets from invitro):
```{r}
input.dir <- "/projects/b1042/GoyalLab/egrody/extractedData/EGS002/singleCell/Seurat/Mmul_10_mac239full/"
SingleCell_umap_env <- read.csv(paste0(input.dir, "invitro_Mmulmac_env.csv"))
SingleCell_umap_env <- SingleCell_umap_env %>% 
  mutate(Count = ifelse(Count == 1, 0, Count)) %>%
  mutate(log1pSingleCell = log1p(Count)) %>% select(-X)

# for reading in Mmul_mac239env aligned
input.dir <- "/projects/b1042/GoyalLab/egrody/extractedData/EGS002/singleCell/counts/Mmul_10_env/"
SingleCell_umap_env <- read.csv(paste0(input.dir, "invitro_Mmulmac_env.csv"))
SingleCell_umap_env <- SingleCell_umap_env %>% 
  rename(log1pSingleCell = env) %>% select(-X)
```


### Loading in SALVEseq reads
Now loading the SALVEseq v2 reads.
```{r}
PL_env <- SeuratSALVEseq("/projects/b1042/GoyalLab/egrody/extractedData/EGS003/counts/count_PLenv_Mmul10env/outs/filtered_feature_bc_matrix/", 
                     "PL_env", 
                     output_dir = "/projects/b1042/GoyalLab/egrody/extractedData/EGS003/Seurat/Mmul_10_env/PL_env/", plots = FALSE)
PL_env_clean <- targetExpressionDF(PL_env, "env")
PL_env_clean <- PL_env_clean %>% filter(env > 0.7) %>%
  rename(log1pSALVE = env)

# to use SVQ instead:
#PL_env <- read.csv("/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/EGS003/scPathoQuant/PL_env/pathogen_al_umi_read_counts_mac239.csv")
#PL_env_clean <- PL_env %>% filter(read > 1) %>% select(-read) %>% rename(cellID = cell_barcode) %>%
#  count(cellID) %>% mutate(cellID = str_sub(cellID, 1, -3)) %>% filter(n>1) %>%
#  mutate(log1pSALVE = log1p(n)) %>% rename(SALVEcount = n)
```

### Linking to 10X
To see what are in common, we use left_join to look at only the cells with 10X anchors:
```{r}
env_paint_umap = left_join(SingleCell_umap_env, PL_env_clean, by = "cellID")
env_paint_umap[is.na(env_paint_umap)] <- 0

left_both <- env_paint_umap %>% filter(log1pSingleCell != 0 & log1pSALVE != 0)
left_onlySingleCell <- env_paint_umap %>% filter(log1pSingleCell != 0)
left_onlySALVE <- env_paint_umap %>% filter(log1pSALVE != 0)
cat("\n\nenv\nTotal cells: \t\t", nrow(env_paint_umap),
  "\nCells in both: \t\t", nrow(left_both),
  "\nCells in 10X only: \t", nrow(left_onlySingleCell) - nrow(left_both),
    "\nCells in SALVE only: \t", nrow(left_onlySALVE) - nrow(left_both))

output_dir <- "/projects/b1042/GoyalLab/egrody/extractedData/EGS003/Seurat/Mmul_10_env/PL_env/"
write.csv(left_both, paste0(output_dir, "left_both.csv"))
write.csv(left_onlySingleCell, paste0(output_dir, "left_onlySingleCell.csv"))
write.csv(left_onlySALVE, paste0(output_dir, "left_onlySALVE.csv"))

```

Now looking at all the SALVE data, even without anchors, using full_join:
```{r}
invitro_paint_umap_full = full_join(SingleCell_umap_env, PL_env_clean, by = "cellID")
invitro_paint_umap_full[is.na(invitro_paint_umap_full)] <- 0

# how many mac239+ cells from each method?
full_both <- invitro_paint_umap_full %>% filter(log1pSingleCell != 0 & log1pSALVE != 0)
full_onlySingleCell <- invitro_paint_umap_full %>% filter(log1pSingleCell != 0)
full_onlySALVE <- invitro_paint_umap_full %>% filter(log1pSALVE != 0)
unanchored <- full_onlySALVE %>% filter(!(cellID %in% left_onlySALVE$cellID))
cat("env\nTotal cells: \t\t", nrow(invitro_paint_umap_full),
  "\nCells in both: \t\t", nrow(full_both),
  "\nCells in 10X only: \t", nrow(full_onlySingleCell) - nrow(full_both),
    "\nCells in SALVE only: \t", nrow(full_onlySALVE) - nrow(full_both),
  "\nUnanchored SALVE: \t", nrow(unanchored))

output_dir <- "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/EGS003/Seurat/Mmul_10_env/PL_env/"
write.csv(full_both, paste0(output_dir, "full_both.csv"))
write.csv(full_onlySingleCell, paste0(output_dir, "full_onlySingleCell.csv"))
write.csv(full_onlySALVE, paste0(output_dir, "full_onlySALVE.csv"))
```

### Correlation plots
```{r}

#  now let's repeat the correlation analysis above
jointTableEnv = inner_join(SingleCell_umap_env, PL_env_clean, by = "cellID") 
jointTableTat = inner_join(SingleCell_umap_tat, PL_tat1_clean, by = "cellID") 

#  mutate(log1pViserCount = log1p(VISERcount), log2ViserCount = log2(VISERcount), lnViserCount = log(VISERcount), normalizedViserCount = 1000*VISERcount/sum(VISERcount), logNormViser = log1p(normalizedViserCount), log1pSingleCell = log1p(SingleCellcount))


output.dir <- "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/EGS003/Seurat/"
cat("Correlation of env between 10X data and log1p SALVE data:", cor(jointTableEnv$log1pSingleCell, jointTableEnv$log1pSALVE))
corrplot <- ggplot(data = jointTableEnv, aes(x=log1pSALVE, y=log1pSingleCell)) +
  geom_point() + 
  xlim(0,7) + 
  ylim(0,7)
corrplot
ggsave(corrplot, file = paste0(output.dir, "PLenv_correlationplot.svg"))

cat("Correlation of tat between 10X data and log1p SALVE data:", cor(jointTableTat$log1pSingleCell, jointTableTat$log1pSALVE))
corrplot <- ggplot(data = jointTableTat, aes(x=log1pSALVE, y=log1pSingleCell)) +
  geom_point() +
  xlim(0,7) +
  ylim(0,7)
ggsave(corrplot, file = paste0(output.dir, "PLtat1_correlationplot.svg"))
```

### Paint UMAPs
```{r}

output.dir <- "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/EGS003/joint/"
umap_env_salve <- create_umap_plot(
  data = env_paint_umap,
  method = "SALVE",
  title = "env from SALVE",
  output_file_name = "umap_env_SALVE.svg",
  output_dir = output.dir
)
create_umap_plot(
  data = tat_paint_umap,
  method = "SALVE",
  title = "tat from SALVE",
  output_file_name = "umap_tat_SALVE.svg",
  output_dir = output.dir
)
create_umap_plot(
  data = env_paint_umap,
  method = "10X",
  title = "env from 10X",
  output_file_name = "umap_env_SingleCell.svg",
  output_dir = output.dir
)
create_umap_plot(
  data = tat_paint_umap,
  method = "10X",
  title = "env from 10X",
  output_file_name = "umap_env_SingleCell.svg",
  output_dir = output.dir
)
```

```{r identity UMAPs}

create_identity_umap(
   data = env_paint_umap,
   title = "env between methods",
   output_file_name = "umap_env_comparison.svg",
   output_dir = output.dir
)

create_identity_umap(
   data = tat_paint_umap,
   title = "tat between methods",
   output_file_name = "umap_tat_comparison.svg",
   output_dir = output.dir
)
```

### Unanchored cells
Do the unanchored cells exist in the raw cellIDs from 10X?
```{r}
input_dir <- "/projects/b1042/GoyalLab/egrody/20231017_EGS002/analysis/counts/Mmul_10_only/run_count_invitro/outs/raw_feature_bc_matrix_bbmap/"
output_dir <- ("/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/EGS003/joint/")
barcodeRankPlot(input_dir, invitro_paint_umap_full, output_dir)

```


## Mmul_10_env (old?)
Redo analyses after aligning to Mmul_10_env

### Compare PL alignment to Mmul+mac vs mac
```{r}
PL_Mmulmac <- SeuratSALVEseq("/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/EGS003/counts/count_PLenv_Mmul10env/outs/filtered_feature_bc_matrix/", 
                     "PL_env", plots = FALSE)
PL_Mmulmac_clean <- targetExpressionDF(PL_Mmulmac, "env")
PL_Mmulmac_clean <- PL_Mmulmac_clean %>% filter(env > 0.7) %>% mutate(cellID = str_sub(cellID, 1, -3)) %>%
  rename(log1pSALVE = env)

PL_mac_data <- Read10X(data.dir = "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/EGS003/counts/count_PLenv_mac239annot/outs/filtered_feature_bc_matrix/")
PL_mac  <- CreateSeuratObject(counts = PL_mac_data, project = "PL_env", min.cells = 3) #no minimum features
PL_mac <- NormalizeData(PL_mac, verbose = FALSE)
PL_mac_clean <- targetExpressionDF(PL_mac, "env")
PL_mac_clean <- PL_mac_clean %>% filter(env > 0.7) %>% mutate(cellID = str_sub(cellID, 1, -3)) %>%
  rename(log1pSALVE = env)

alignments_yield <- full_join(PL_Mmulmac_clean, PL_mac_clean, by = "cellID") %>%
  rename(log1pmac = log1pSALVE.y, log1pMmulmac = log1pSALVE.x)
alignments_yield[is.na(alignments_yield)] <- 0
alignments_both <- alignments_yield %>% filter(log1pMmulmac != 0 & log1pmac != 0)
alignments_onlymac <- alignments_yield %>% filter(log1pmac != 0)
alignments_onlyMmul <- alignments_yield %>% filter(log1pMmulmac != 0)
cat("PL alignments\nTotal cells: \t\t", nrow(alignments_yield),
  "\nCells in both: \t\t", nrow(alignments_both),
  "\nCells in Mmulmac only: \t", nrow(alignments_onlyMmul) - nrow(alignments_both),
    "\nCells in mac only: \t", nrow(alignments_onlymac) - nrow(alignments_both))

```


### env analysis


## bamsort

### bam_sorted vs SVQ (old)
Reminder that the 10X results are from scViralQuant, not scPathoQuant
```{r}
SVQ <- read.csv("/projects/b1042/GoyalLab/egrody/20231017_EGS002/analysis/scViralQuant/invitro/gtf/virus_al_gene_counts.csv")
bamsort <- read.csv("/projects/b1042/GoyalLab/egrody/20231017_EGS002/analysis/bam_sorted/invitro_bam_dedup_subset_env.csv")

SVQ_env <- SVQ %>% filter(str_detect(gene, "env|nef|UTR")) %>% 
  group_by(cell_barcode) %>%
  summarise(total_umi = sum(umi)) %>%
  mutate(logcount = log1p(total_umi))

bamsort <- bamsort %>% mutate(logcount = log1p(Read_Count)) %>% 
  rename(cell_barcode = Cell_ID)

df <- data.frame(
  value = c(SVQ_env$logcount, bamsort$logcount),
  group = rep(c("SVQ", "bamsort"), c(length(SVQ_env$logcount), length(bamsort$logcount)))
)

# Create the plot
ggplot(df, aes(x = value, fill = group)) +
  geom_histogram(position = "identity", alpha = 0.5, bins = 30) +
  scale_fill_manual(values = c("SVQ" = "red", "bamsort" = "blue")) +
  labs(title = "Overlapping Histograms", x = "Value", y = "Count") +
  theme_minimal()


jointTable = full_join(SVQ_env, bamsort, by = "cell_barcode")
jointTable[is.na(jointTable)] <- 0

#output.dir <- "/projects/b1042/GoyalLab/egrody/20240116_EGS003_004/EGS003/Seurat/"
cat("Correlation of env between SVQ and bamsort:", cor(jointTable$logcount.x, jointTable$logcount.y))
corrplot <- ggplot(data = jointTable, aes(x=logcount.x, y=logcount.y)) +
  geom_point() + 
  xlim(0,7) + 
  ylim(0,7) +
  xlab("SVQ log counts") + 
  ylab("bamsort log counts")
corrplot
ggsave(corrplot, file = paste0(output.dir, "bamsortSVQ_correlation.svg"))
```

### Filter by 10X cellIDs (old)
```{r}
SingleCell_cells <- Cells(invitro_min200)
bamsort_filter <- bamsort %>%
  filter(cell_barcode %in% SingleCell_cells)
SVQ_filter <- SVQ_env %>%
  filter(cell_barcode %in% SingleCell_cells)
```

### envupstream vs envdownstream vs SALVE (updated)
When we did the joint analysis, there were >hundred cells that were env+ only in 10X. Is this because their env reads were upstream of our SALVE primer?
```{r}
input_dir <- "/projects/b1042/GoyalLab/egrody/extractedData/EGS002/singleCell/bamsort/alignment/"
upstream <- read.csv(paste0(input_dir, "invitro_bamsort_env_upstream.csv")) %>% 
  unique() %>%
  count(Cell_ID) %>%
  rename(Count = n, cellID = Cell_ID) %>%
  filter(Count > 1) 

#older, skip
downstream <- read.csv(paste0(input_dir, "invitro_bamsort_env_downstream.csv")) %>% 
  filter(Read_Count > 1) %>%
  rename(cellID = Cell_ID) %>%
  mutate(cellID = str_sub(cellID, end=-3))
downstream_any <- read.csv(paste0(input_dir, "invitro_bam_dedup_subset_env.csv")) %>% 
  filter(Read_Count > 1) %>%
  rename(cellID = Cell_ID) %>%
  mutate(cellID = str_sub(cellID, end=-3))
upstream_any <- read.csv(paste0(input_dir, "invitro_bamsort_env_upstream_any.csv")) %>% 
  filter(Read_Count > 1) %>%
  rename(cellID = Cell_ID) %>%
  mutate(cellID = str_sub(cellID, end=-3))

#restart here
upstream_valid <- upstream %>%
  filter(cellID %in% env_paint_umap$cellID)
downstream_valid <- downstream %>%
  filter(cellID %in% env_paint_umap$cellID)
downstream_any_valid <- downstream_any %>%
  filter(cellID %in% env_paint_umap$cellID)
upstream_any_valid <- upstream_any %>%
  filter(cellID %in% env_paint_umap$cellID)

upstream_only <- upstream_valid %>%
  mutate(common = cellID %in% downstream$cellID) %>%
  filter(common == FALSE)
downstream_only <- downstream_valid %>%
  mutate(common = cellID %in% upstream$cellID) %>%
  filter(common == FALSE)
upstream_any_only <- upstream_any_valid %>%
  mutate(common = cellID %in% downstream_any$cellID) %>%
  filter(common == FALSE)

allEnv <- bind_rows(upstream, downstream, downstream_any) %>%
  distinct()
allEnv_valid <- bind_rows(upstream_valid, downstream_valid, downstream_any_valid) %>%
  distinct()

either <- invitro_paint_umap_full$cellID[invitro_paint_umap_full$log1pSingleCell > 0 | invitro_paint_umap_full$log1pSALVE > 0]
both <- invitro_paint_umap_full$cellID[invitro_paint_umap_full$log1pSingleCell > 0 & invitro_paint_umap_full$log1pSALVE > 0]
SingleCell_only <- invitro_paint_umap_full$cellID[invitro_paint_umap_full$log1pSingleCell > 0 & invitro_paint_umap_full$log1pSALVE == 0]
SALVE_only <- invitro_paint_umap_full$cellID[invitro_paint_umap_full$log1pSingleCell == 0 & invitro_paint_umap_full$log1pSALVE > 0]


cat(
  "# cells from bamsort total: \t\t",
  nrow(unique(allEnv)),
  "\n# cells from bamsort total in 10X: \t",
  nrow(unique(allEnv_valid)),
  "\n# cells in upstream_valid: \t\t",
  length(upstream_valid$cellID),
  "\n# cells in downstream_valid: \t\t",
  length(downstream_valid$cellID),
  "\n# cells in upstream_only: \t\t",
  length(upstream_only$cellID),
  "\n# cells in downstream_only: \t\t",
  length(downstream_only$cellID),
  "\n\n# cells in 10X only: \t\t\t",
  length(SingleCell_only),
  "\n# cells in SALVE only: \t\t\t",
  length(SALVE_only),
  "\n# cells in both: \t\t\t",
  length(both),
  "\n\n# cells in upstream in both: \t\t",
  sum(upstream$cellID %in% both),
  "\n# cells in upstream in only SALVE: \t",
  sum(upstream$cellID %in% SALVE_only),
  "\n# cells in upstream in only 10X: \t",
  sum(upstream$cellID %in% SingleCell_only),
  "\n\n# cells in downstream in both: \t\t",
  sum(downstream$cellID %in% both),
  "\n# cells in downstream in only SALVE: \t",
  sum(downstream$cellID %in% SALVE_only),
  "\n# cells in downstream in only 10X: \t",
  sum(downstream$cellID %in% SingleCell_only),
  "\n\n# cells in downstream_any in both: \t",
  sum(downstream_any$cellID %in% both),
  "\n# cells in downstream_any in only SALVE: \t",
  sum(downstream_any$cellID %in% SALVE_only),
  "\n# cells in downstream_any in only 10X: \t",
  sum(downstream_any$cellID %in% SingleCell_only),
  "\n\n# cells in all bamsort in both: \t",
  sum(unique(allEnv_valid$cellID) %in% both),
  "\n# cells in all bamsort in only SALVE: \t",
  sum(unique(allEnv$cellID) %in% SALVE_only),
  "\n# cells in all bamsort in only 10X: \t",
  sum(unique(allEnv_valid$cellID) %in% SingleCell_only),
  "\n\n# cells in upstream_only in both: \t",
  sum(upstream_only$cellID %in% both),
  "\n# cells in upstream_only in only SALVE: \t",
  sum(upstream_only$cellID %in% SALVE_only),
  "\n# cells in upstream_only in only 10X: \t",
  sum(upstream_only$cellID %in% SingleCell_only),
  "\n\n# cells in upstream_any_only in both: \t",
  sum(upstream_any_only$cellID %in% both),
  "\n# cells in upstream_any_only in only SALVE: \t",
  sum(upstream_any_only$cellID %in% SALVE_only),
  "\n# cells in upstream_any_only in only 10X: \t",
  sum(upstream_any_only$cellID %in% SingleCell_only)
)

cat(
  "# cells from bamsort upstream: \t\t",
  nrow(unique(upstream)),
  "\n# cells from bamsort upstream in 10X: \t",
  nrow(unique(upstream_valid)),
  "\n\n# cells in 10X only: \t\t\t",
  length(SingleCell_only),
  "\n# cells in SALVE only: \t\t\t",
  length(SALVE_only),
  "\n# cells in both: \t\t\t",
  length(both),
  "\n\n# cells in upstream in both: \t\t",
  sum(upstream_valid$cellID %in% both),
  "\n# cells in upstream in only SALVE: \t",
  sum(upstream_valid$cellID %in% SALVE_only),
  "\n# cells in upstream in only 10X: \t",
  sum(upstream_valid$cellID %in% SingleCell_only)
)


```



### 10X env+ only



## Subset analysis

```{r}
output_dir <- "/projects/b1042/GoyalLab/egrody/extractedData/EGS003/EGS002/Mmul_10_only/invitro/plots/"
invitro <- SeuratPipeline("/projects/b1042/GoyalLab/egrody/extractedData/EGS003/EGS002/Mmul_10_only/invitro/filtered_feature_bc_matrix/", "invitro")
dplot <- DimPlot(invitro, reduction = "umap")
ggsave(dplot, file = paste0(output_dir, "invitro_umap.svg"))
invitro_coords <- export_cell_metadata(invitro,                            "/projects/b1042/GoyalLab/egrody/extractedData/EGS003/EGS002/invitro_umapclusters.csv")

output.dir <- "/projects/b1042/GoyalLab/egrody/20231017_EGS002/analysis/scViralQuant/outputs/"
SingleCell_invitro_umap <- read.csv(paste0(output.dir, "invitro_SingleCellumap_env.csv"))
```


### VISER
```{r}
output.dir <- "/projects/b1042/GoyalLab/egrody/20230929_VISER/analysis/seqIOdualRead/splits/cleaned/"
VISER_invitro_whitelist <- read.csv(paste0(output.dir, "invitro_VISER_whitelist_uniqued.csv")) %>% select(-X)
VISER_invitro_whitelist <- VISER_invitro_whitelist %>%
  mutate(cellID = paste0(cellID, "-1")) %>%
  filter(n > 1)
SingleCell_invitro <- SingleCell_invitro_umap %>%
  mutate(cellID = paste0(cellID, "-1")) %>%
  select(cellID, SingleCellcount) %>%
  filter(SingleCellcount > 1)
test_invitro_paint_umap = left_join(invitro_coords, VISER_invitro_whitelist,
                                    by = "cellID") %>% rename(VISERcount = n) %>% 
                    mutate(log2ViserCount = log2(VISERcount))
invitro_paint_umap = left_join(test_invitro_paint_umap, SingleCell_invitro,
                                    by = "cellID") %>% 
                    mutate(log2SingleCell = log2(SingleCellcount))
invitro_paint_umap[is.na(invitro_paint_umap)] <- 0

```

### SALVE
```{r}
invitro_coords <- read.csv("/projects/b1042/GoyalLab/egrody/extractedData/EGS003/EGS002/invitro_umapclusters.csv")

output.dir <- "/projects/b1042/GoyalLab/egrody/extractedData/EGS003/Seurat/"
SALVE_umap_env <- read.csv(paste0(output.dir, "PL_env/PLenv_mac239expression.csv"))
SALVE_umap_tat <- read.csv(paste0(output.dir, "PL_tat1/PLtat1_mac239expression.csv"))
SALVE_invitro <- SALVE_umap_env %>%
  mutate(cellID = paste0(cellID, "-1")) %>%
  select(-X) %>%
  filter(Count > 1)
SingleCell_invitro <- SingleCell_invitro_umap %>%
  mutate(cellID = paste0(cellID, "-1")) %>%
  select(cellID, SingleCellcount) %>%
  filter(SingleCellcount > 1)
test_invitro_paint_umap = left_join(invitro_coords, SALVE_invitro,
                                    by = "cellID")
invitro_paint_umap = left_join(test_invitro_paint_umap, SingleCell_invitro,
                                    by = "cellID") %>% 
                    mutate(log2SingleCell = log2(SingleCellcount))
invitro_paint_umap[is.na(invitro_paint_umap)] <- 0


```



Quick plotting things
```{r}

# Function to create Seurat-style DimPlot
plot_clusters <- function(data, 
                         label_clusters = TRUE,
                         pt_size = 0.5,
                         label_size = 4) {
  
  # Create custom color palette similar to Seurat
  n_clusters <- length(unique(data$Cluster))
  cluster_colors <- scales::hue_pal()(n_clusters)
  
  # Create plot
  p1 <- ggplot(data, aes(x = UMAP1, y = UMAP2, color = Cluster)) +
    geom_point(size = pt_size) +
    scale_color_manual(values = cluster_colors) +
    theme_bw() +
    theme(
      panel.grid = element_blank(),
      axis.title = element_text(size = 12),
      legend.title = element_blank(),
      panel.border = element_rect(colour = "black", fill = NA)
    )
  
  # Add cluster labels if requested
  if (label_clusters) {
    # Calculate cluster centers
    cluster_centers <- data %>%
      group_by(Cluster) %>%
      summarise(
        UMAP1 = median(UMAP1),
        UMAP2 = median(UMAP2)
      )
    
    p1 <- p1 + geom_text(data = cluster_centers,
                         aes(label = Cluster),
                         size = label_size,
                         color = "black")
  }
  
  return(p1)
}

# Function to create Seurat-style FeaturePlot
plot_gene_expression <- function(data,
                               countcolumn,
                               pt_size = 0.5,
                               min_cutoff = NA,
                               max_cutoff = NA) {
  
  # Create copy of data for potential expression cutoffs
  plot_data <- data
  
  # Apply cutoffs if specified
  if (!is.na(min_cutoff)) {
    plot_data[[countcolumn]][plot_data[[countcolumn]] < min_cutoff] <- min_cutoff
  }
  if (!is.na(max_cutoff)) {
    plot_data[[countcolumn]][plot_data[[countcolumn]] > max_cutoff] <- max_cutoff
  }
  
  # Create plot
  p2 <- ggplot(plot_data, aes(x = UMAP1, y = UMAP2)) +
    # Plot cells with no expression first
    geom_point(data = subset(plot_data, plot_data[[countcolumn]] <= 0),
               color = "gray93",
               size = pt_size) +
    # Plot expressing cells on top
    geom_point(data = subset(plot_data, plot_data[[countcolumn]] > 0),
               aes_string(color = countcolumn),
               size = pt_size) +
    scale_color_gradient(low = "gray93", high = "darkblue",
                        name = "Expression") +
    theme_bw() +
    theme(
      panel.grid = element_blank(),
      axis.title = element_text(size = 12),
      panel.border = element_rect(colour = "black", fill = NA)
    )
  
  return(p2)
}

cluster_plot <- plot_clusters(test_invitro_paint_umap)
xprn_plot_SALVE <- plot_gene_expression(invitro_paint_umap, "Count")
xprn_plot_SingleCell <- plot_gene_expression(invitro_paint_umap, "log2SingleCell")
intersect_umap <- invitro_paint_umap %>%
  mutate(Count = ifelse(log2SingleCell == 0, 0, Count))
xprn_plot_intersect <- plot_gene_expression(intersect_umap, "Count") #SALVE xprn in cells that are 10X+

output_dir <- "/projects/b1042/GoyalLab/egrody/extractedData/EGS003/joint/Mmul_10_onlyclusters/"
ggsave(cluster_plot, file = paste0(output_dir, "invitro_umap_cluster.svg"))
ggsave(xprn_plot_SALVE, file = paste0(output_dir, "invitro_umap_SALVE.svg"))
ggsave(xprn_plot_SingleCell, file = paste0(output_dir, "invitro_umap_SingleCell.svg"))
ggsave(xprn_plot_intersect, file = paste0(output_dir, "invitro_umap_intersect.svg"))

cat("SingleCell vRNA+: ", sum(invitro_paint_umap$log2SingleCell != 0, na.rm = TRUE),
    "\nSALVE vRNA+: ", sum(invitro_paint_umap$Count != 0, na.rm = TRUE),
    "\nCells in common: ", sum(intersect_umap$Count != 0, na.rm = TRUE))

write.csv(invitro_paint_umap, paste0(output_dir, "invitro_paint_umap.csv"))


```

Quick correlation plot:
```{r}
innerjoin <- invitro_paint_umap %>%
  mutate(log2SingleCell = ifelse(Count == 0, NA, log2SingleCell)) %>%
  mutate(Count = ifelse(log2SingleCell == 0, 0, Count)) %>%
  filter(!is.na(log2SingleCell)) %>% filter(!is.na(Count))

innerjoin <- invitro_paint_umap %>%
  mutate(log2SingleCell = ifelse(Count == 0, NA, log2SingleCell)) %>%
  mutate(Count = ifelse(log2SingleCell == 0, 0, Count)) %>%
  filter(log2SingleCell != 0)

cat("Correlation of invitro between 10X and SALVE data:", cor(innerjoin$log2SingleCell, innerjoin$Count))
corrplot <- ggplot(data = innerjoin, aes(x=Count, y=log2SingleCell)) +
  geom_point() + 
  xlim(0,10) + 
  ylim(0,10)
corrplot
ggsave(corrplot, file = paste0(output_dir, "intersect_correlationplot.svg"))
```



Let's pull the list of cellIDs that are vRNA positive (just from 10X itself (SPQ)).
```{r}
input.dir <- "/projects/b1042/GoyalLab/egrody/extractedData/EGS003/EGS002/"
SingleCell_umap_env <- read.csv(paste0(input.dir, "invitro_scVQexpression_env.csv"))
SingleCell_umap_env <- SingleCell_umap_env %>% mutate(log1pSingleCell = log1p(Count)) %>% select(-X)

##cheap way out:
list_of_cellIDs <- SingleCell_umap_env %>% filter(log1pSingleCell >= 1) %>%
  select(cellID) %>% 
  mutate(cellID = paste0(cellID, "-1"))
list_of_cellIDs <- list_of_cellIDs$cellID

##skip for now
#invitro_paint_umap_full = full_join(SingleCell_umap_env, PL_env, by = "cellID")
#invitro_paint_umap_full[is.na(invitro_paint_umap_full)] <- 0
#SALVE_cellids <- invitro_paint_umap_full %>% filter(log1pSALVE > 0) %>% select(cellID) %>% mutate(cellID = paste0(cellID, "-1"))
```
Or use above's "SingleCell_invitro"

### DEG analysis

```{r}
# Create your subset of cells
list_of_cells <- invitro_paint_umap %>% filter(invitro_paint_umap$Count != 0) #for SALVE
#list_of_cells <- invitro_paint_umap %>% filter(invitro_paint_umap$SingleCellcount != 0) #for 10X
list_of_cells <- list_of_cells$cellID
#invitro <- readRDS(file = paste0("/projects/b1042/GoyalLab/egrody/extractedData/EGS003/EGS002/", "invitro.rds")) #reload invitro from RDS
subset_cells <- subset(invitro, cells = list_of_cells) 


# Run analysis
results <- analyze_cell_subset(invitro, list_of_cells)
```

### Saving results
```{r}

output.dir <- "/projects/b1042/GoyalLab/egrody/extractedData/EGS003/joint/Mmul_10_onlyclusters/DEG_SALVE/"
save_subset_analysis(results, output_dir = output.dir)

output.dir <- "/projects/b1042/GoyalLab/egrody/extractedData/EGS003/joint/"
plot_subset_analysis(invitro, results, list_of_cells, output_dir = paste0(output.dir, "plots/"))
```

### Interpreting results
```{r}
# Create filtered results list
filteredDEG <- lapply(results$cluster_specific, function(df) {
  if (!is.null(df) && nrow(df) > 0) {
    # Filter for significant adjusted p-value and log2FC threshold
    df[df$p_val_adj < 0.05 & abs(df$avg_log2FC) > 0.5, ]
  } else {
    NULL
  }
})

# Remove any empty results
filteredDEG <- filteredDEG[sapply(filteredDEG, function(x) !is.null(x) && nrow(x) > 0)]

# Print summary of how many genes passed filters in each cluster
for (cluster in names(filteredDEG)) {
  cat(sprintf("Cluster %s: %d genes\n", cluster, nrow(filteredDEG[[cluster]])))
}

# Saving
wb <- createWorkbook()
# Add each cluster's results as a separate worksheet
for (cluster_name in names(filteredDEG)) {
    de_results <- filteredDEG[[cluster_name]]
    if (!is.null(de_results) && nrow(de_results) > 0) {
        # Add cluster results as a worksheet
        addWorksheet(wb, cluster_name)
        # Add data with gene names as a column
        de_results$gene <- rownames(de_results)
        writeData(wb, cluster_name, de_results)
    }
}

# Save the workbook
saveWorkbook(wb, paste0(output.dir, "cluster_specific_DE_filtered.xlsx"), overwrite = TRUE)



filteredConserv <- results$conserved_markers %>%
    filter(p_val_adj < 0.05 & abs(avg_log2FC) > 0.5)

# Print summary of how many genes passed filters in each cluster
cat(sprintf("Conserved markers: %d genes\n", nrow(filteredConserv)))

# For each gene in conserved_markers, let's check which clusters show it as significant
check_gene_presence <- function(gene, cluster_results, p_val_thresh = 0.05, log2fc_thresh = 0.5) {
  # Initialize list to store results
  presence <- list()
  
  # Check each cluster
  for(cluster in names(cluster_results)) {
    de_results <- cluster_results[[cluster]]
    if(!is.null(de_results) && gene %in% rownames(de_results)) {
      # Get gene stats for this cluster
      gene_stats <- de_results[gene,]
      # Check if it meets significance criteria
      if(gene_stats$p_val_adj < p_val_thresh && abs(gene_stats$avg_log2FC) > log2fc_thresh) {
        presence[[cluster]] <- c(
          p_val_adj = gene_stats$p_val_adj,
          avg_log2FC = gene_stats$avg_log2FC
        )
      }
    }
  }
  
  # Return number of clusters and which clusters
  return(list(
    n_clusters = length(presence),
    clusters = names(presence),
    details = presence
  ))
}

# First get cluster counts for each gene
cluster_counts <- lapply(rownames(filteredConserv), function(gene) {
  result <- check_gene_presence(gene, results$cluster_specific)
  data.frame(
    gene = gene,
    num_clusters = result$n_clusters,
    clusters = paste(result$clusters, collapse=",")
  )
}) %>% bind_rows()

# Add cluster information to filtered results
filteredConserv <- filteredConserv %>%
  mutate(gene = rownames(.)) %>%
  left_join(cluster_counts, by = "gene") %>%
  arrange(desc(num_clusters), desc(abs(avg_log2FC)))

# Save to Excel
wb <- createWorkbook()
addWorksheet(wb, "filtered_conserved")
writeData(wb, "filtered_conserved", filteredConserv)
saveWorkbook(wb, paste0(output.dir, "conserved_markers_filtered.xlsx"), overwrite = TRUE)

```



## barcode rank plot

```{r}
input_dir <- "/projects/b1042/GoyalLab/egrody/20231017_EGS002/analysis/counts/Mmul_10_only/run_count_invitro/outs/raw_feature_bc_matrix_bbmap/"
output_dir <- ("/projects/b1042/GoyalLab/egrody/extractedData/EGS003/joint/")
barcodeRankPlot(input_dir, invitro_paint_umap_full, output_dir)
```
