---
title: "EGS001_Mmu"
author: "Emanuelle Grody"
date: "2023-05-31"
output: html_document
---


If installing Seurat for the first time, you'll need to delete and reinstall the R package Matrix. Go to the library location of R and manually delete the Matrix folder in order to remove the old version. To find where the packages are stored, use *.libPaths()* into the console then copy the result as the path and deleted the folder Matrix. Then use *install.packages("Matrix")* to get the correct version.

```{r, echo = FALSE}
source("~/SALVEseq/packages.R")
source("~/SALVEseq/functions.R")
```

# Single Cell Data

If you'd like, you can skip down to the section "Current Analysis"

## Previous analysis
### Seurat whole pipeline

```{r echo = FALSE}
# Load the CD4 dataset
read.dir <- "/projects/b1042/GoyalLab/egrody/20230424_EGS001/singleCell/counts/filtered_feature_bc_matrix/"
output.dir = "/projects/b1042/GoyalLab/egrody/20230424_EGS001/singleCell/analysis/plots/"
mmu <- SeuratPipeline(read.dir, "mmu", output_dir = output.dir, plots = TRUE, rds = TRUE)
```

### Loading in RDS

```{r}
mmu <- readRDS("/projects/b1042/GoyalLab/egrody/20230424_EGS001/singleCell/analysis/mmu.rds")
```


### KLRB1 analysis
KLRB1 expression in UMAP:
```{r}
# Adjust the contrast in the plot
fplot <- FeaturePlot(mmu, features = "KLRB1", min.cutoff = 1, max.cutoff = 3)
ggsave(fplot, file = paste0(output.dir, "mmu_KLRB1_featureplot.svg"), width = 6, height = 6)
```

The default is to have a min cutoff of 1 and a max cutoff of 3 for feature plot. I confirmed that this was appropriate for our sample by plotting in a histogram all cells that have KLRB1 expression and looking for a population cutoff.

To isolate out the KLRB1 expression data:
```{r save UMAP csv}
SingleCell_KLRB1_umap <- targetExpressionDF(mmu, "KLRB1")
write.csv(SingleCell_KLRB1_umap, paste0(output.dir, "KLRB1_SingleCellUMAP.csv"))
```


## Current analysis

```{r}
output.dir <- "/projects/b1042/GoyalLab/egrody/20230424_EGS001/singleCell/analysis/"
# this dataframe includes zeros
SingleCell_KLRB1 <- read.csv(paste0(output.dir, "KLRB1_SingleCellUMAP.csv"))
#mmu <- readRDS(paste0(output.dir, "mmu.rds"))
```


# VISER analysis

After subsetting my Read1 and Read2 fastq files to 1M reads, I used a Python script to filter and isolate out the high quality KLRB1 reads from the subset.

## Previous analysis (skip)

### Target cutoff

Previously, I calculated the Levenshtein distance (must keep target column for this):
```{r, eval = FALSE}
output.dir <- "/projects/b1042/GoyalLab/egrody/20230424_EGS001/VISER/analysis/KLRB1/"
VISER_KLRB1 <- read.csv(paste0(output.dir, "VISER_KLRB1_full.csv"))

expectedTarget = "AAAGTTCTTCACCTTCATCTCTTCCTCGGGATGTCTG" 
VISER_KLRB1$dist <- NA
for (i in 1:nrow(VISER_KLRB1)) {
  # Calculate the Levenshtein distance between the current string and the constant
  dist <- adist(VISER_KLRB1$target[i], expectedTarget)
  # Assign the distance to the 'dist' column
  VISER_KLRB1$dist[i] <- dist
}

hist(recovered_KLRB1$dist, breaks = 50, xlab = "Levenshtein Distance: recovered to expected")
```
Early in my analysis, I looked at this histogram to determine that a good cutoff for Levenshtein distance is 6. The data here has already been trimmed to this Levenshtein distance (cutoff set in Python seqIO script).


## Current analysis

After running seqIO and postSeqIO Python scripts, load in the clean data here.
```{r}
output.dir = "/projects/b1042/GoyalLab/egrody/20230424_EGS001/VISER/analysis/KLRB1/postSeqIO/"
VISER_KLRB1_clean <- read.csv(paste0(output.dir, "VISER_KLRB1_clean.csv")) %>%
  mutate(cellID = paste0(cellID, "-1")) %>%
  rename(VISERcount = count)

output.dir = "/projects/b1042/GoyalLab/egrody/20230424_EGS001/singleCell/analysis/"
SingleCell_KLRB1 <- read.csv(paste0(output.dir, "KLRB1_SingleCellUMAP.csv")) %>% select(-X) %>%
  rename(SingleCellcount = KLRB1)
```


### Joint UMAPs
Because VISER data is all non-zero (unlike 10X), I use log2 (binary log) instead of using log1p (natural log plus 1), which is what 10X does.

This is to plot the VISER cellIDs on the 10X UMAP.
```{r}
left_paint_umap = left_join(SingleCell_KLRB1, VISER_KLRB1_clean, by = "cellID") %>% 
     mutate(log2ViserCount = log2(VISERcount)) %>%
  mutate(log1pSingleCell = log1p(SingleCellcount))
left_paint_umap[is.na(left_paint_umap)] <- 0
output.dir <- "/projects/b1042/GoyalLab/egrody/20230424_EGS001/VISER/analysis/KLRB1/joint/"
plotUMAP(left_paint_umap, log2ViserCount, "KLRB1 from VISER", output.dir, "UMAP_VISERplot.svg")
plotUMAP(left_paint_umap, log1pSingleCell, "KLRB1 from 10X", output.dir, "UMAP_SingleCellplot.svg")
```
Now normalizing the UMAPs and dividing one by the other to see the relationships between the specific cellIDs:
```{r}
full_paint_umap = full_join(SingleCell_KLRB1, VISER_KLRB1_clean, by = "cellID") %>% 
     mutate(log2ViserCount = log2(VISERcount)) %>%
  mutate(log1pSingleCell = log1p(SingleCellcount)) 
full_paint_umap[is.na(full_paint_umap)] <- 0
full_paint_umap <- full_paint_umap %>% 
  mutate(normSingleCell = log1pSingleCell/max(log1pSingleCell), normlogVISER = log2ViserCount/max(log2ViserCount)) %>%
  mutate(ratioSingleVISER = (normSingleCell+1)/(normlogVISER+1))

plotUMAP(full_paint_umap, ratioSingleVISER, "Counts Ratio SingleCell:VISER", output.dir, "UMAP_ratio.svg", comparison = TRUE)
```

Identities UMAP:
```{r}
create_identity_umap(left_paint_umap, "KLRB1 between methods", "UMAP_identities.svg", output.dir)
```

### Venn diagram numbers
```{r}
left_both <- sum((left_paint_umap$SingleCellcount != 0 & left_paint_umap$log2ViserCount != 0), na.rm = TRUE)
left_10Xonly <- sum(left_paint_umap$SingleCellcount != 0, na.rm = TRUE) - left_both
left_SALVEonly <- sum(left_paint_umap$log2ViserCount != 0, na.rm = TRUE) - left_both
cat("KLRB1\nTotal cells: \t\t", nrow(left_paint_umap),
  "\nCells in both: \t\t", left_both,
  "\nCells in 10X only: \t", left_10Xonly,
    "\nCells in VISER only: \t", left_SALVEonly)
```


### Correlation plots
```{r}
results <- correlation_plots(VISER_KLRB1_clean, SingleCell_KLRB1, join = "inner")

# head(results$data)

# # View individual plots
# results$plots$SALVEcount
# results$plots$log1pSalveCount
 
# # Print all plots at once
library(gridExtra)
do.call(grid.arrange, c(results$plots, ncol = 2))


#saving
ggsave(filename = paste0(output.dir, "correlation_countsvscounts.svg"), results$plots$SALVEcount, 
       width = 7, height = 6)
ggsave(filename = paste0(output.dir, "correlation_countsvslog2SALVE.svg"), results$plots$log2SalveCount, 
       width = 7, height = 6)

```


### Cells High in VISER and zero in 10X
```{r}
# finding out which cells are the ones that are on the X axis in the above correlation plots
whoareyou <- jointTable %>% filter(SingleCellcount == 0) %>% select(cellID, logNormViser) %>% rename(onlyinViser = logNormViser)
#paint_umap <- mutate(whoisthey = whoareyou)
whoiswhere <- full_join(whoareyou, paint_umap, by = "cellID") %>% select(onlyinViser, UMAP1, UMAP2)
whoiswhere[is.na(whoiswhere)] <- 0

output.dir <- "/projects/b1042/GoyalLab/egrody/20230424_EGS001/VISER/analysis/KLRB1/joint/"
plotUMAP(whoiswhere, onlyinViser, "KLRB1 from VISER", output.dir, "UMAP_whoiswhere.svg")
```

How much RNA overall do these cells have? Maybe they just have less total RNA
```{r}
DefaultAssay(mmu) <- "RNA"
test <- data.frame(cellID = Cells(mmu))
whoareyou <- whoareyou %>% mutate(cellID = paste0(cellID, "-1")) #remember, this is how the cellIDs are stored in the Seurat object
whoareyoujoin <- inner_join(test, whoareyou, by = "cellID")

subset_whoareyou <- subset(mmu, cells = whoareyoujoin$cellID)

VlnPlot(subset_whoareyou, features = c("nFeature_RNA", "nCount_RNA", "percent_mito"), ncol = 3)
VlnPlot(mmu, features = c("nFeature_RNA", "nCount_RNA", "percent_mito"), ncol = 3)
```
After rerunning the first section to reload in the raw counts into mmu (not normalized), I want to check that these cells really have no KLRB1 expression.
```{r}
subset_whoareyou_noNorm <- subset(mmu, cells = whoareyoujoin$cellID)
noNorm <- GetAssayData(object = subset_whoareyou_noNorm, assay = "RNA", slot = "data")["KLRB1",]
plot(noNorm)
max(noNorm)
```


### Barcode rank plot
```{r}
raw_location = "/projects/b1042/GoyalLab/egrody/20230424_EGS001/singleCell/counts/raw_feature_bc_matrix/"

bcrp <- read.csv( "/projects/b1042/GoyalLab/egrody/20230424_EGS001/VISER/analysis/KLRB1/postSeqIO/VISER_KLRB1_clean.csv")
bcrp <- bcrp %>% 
  mutate(log1pSALVE = log1p(count))

barcodeRankPlot(rawDataFolder = raw_location,
                jointFullJoin = bcrp,
                output_dir = "/projects/b1042/GoyalLab/egrody/20230424_EGS001/VISER/analysis/KLRB1/joint/",
                plotTitle = "Barcode Rank Plot: EGS001 KLRB1 SALVE v1")
```


